File: packages/brand-rules/src/context.ts
import type { JsonValue } from "../../design-ir/src/index";
import type { BrandAssetSet, BrandContext, BrandRuleSet, BrandTokenSet } from "./types";
import { BrandRuleError, validateBrandRuleSet } from "./runtime";

function selectedTokens(tokenSet: BrandTokenSet): Readonly<Record<string, JsonValue>> {
  const colors: Record<string, JsonValue> = {};
  for (const token of [...tokenSet.colors].sort((a, b) => a.id.localeCompare(b.id))) {
    colors[token.id] = { name: token.name, value: token.value, roles: [...token.roles] };
  }
  const fonts: Record<string, JsonValue> = {};
  for (const token of [...tokenSet.fonts].sort((a, b) => a.id.localeCompare(b.id))) {
    fonts[token.id] = {
      name: token.name,
      asset_id: token.asset_id,
      roles: [...token.roles],
      fallbacks: [...(token.fallback_asset_ids ?? [])],
    };
  }
  return {
    colors,
    fonts,
    spacing_scale: [...tokenSet.spacing_scale],
  };
}

export function buildBrandContext(
  ruleSet: BrandRuleSet,
  tokenSet: BrandTokenSet,
  assetSet: BrandAssetSet,
): BrandContext {
  validateBrandRuleSet(ruleSet);
  if (ruleSet.status !== "PUBLISHED") throw new BrandRuleError("BrandContext requires a PUBLISHED BrandRuleSet");
  if (tokenSet.brand_profile_id !== ruleSet.brand_profile_id || assetSet.brand_profile_id !== ruleSet.brand_profile_id) {
    throw new BrandRuleError("BrandContext brand profile mismatch");
  }
  if (tokenSet.version !== ruleSet.token_set_version || assetSet.version !== ruleSet.asset_set_version) {
    throw new BrandRuleError("BrandContext version mismatch");
  }

  const hardRules = ruleSet.rules
    .filter((rule) => rule.active && rule.severity === "HARD")
    .sort((a, b) => b.priority - a.priority || a.id.localeCompare(b.id));
  const allowedAssets = [...new Set([
    ...assetSet.logo_asset_ids,
    ...assetSet.font_asset_ids,
    ...assetSet.reference_asset_ids,
  ])].sort();

  return {
    brand_profile_id: ruleSet.brand_profile_id,
    brand_rule_set_id: ruleSet.id,
    brand_rule_set_version: ruleSet.version,
    hard_rules: hardRules,
    selected_tokens: selectedTokens(tokenSet),
    allowed_assets: allowedAssets,
    voice_summary: {
      tone_attributes: [...ruleSet.voice.tone_attributes],
      preferred_vocabulary: [...ruleSet.voice.preferred_vocabulary],
      forbidden_terms: [...ruleSet.voice.forbidden_terms],
    },
    reference_asset_ids: [...ruleSet.visual_references.reference_asset_ids].sort(),
    pinned: true,
  };
}

File: services/brand-rules/src/lumi_brand_rules/context.py
from __future__ import annotations

from typing import Any

from .model import BrandAssetSet, BrandRuleError, BrandRuleSet, BrandTokenSet, validate_rule_set


def build_brand_context(
    rule_set: BrandRuleSet,
    token_set: BrandTokenSet,
    asset_set: BrandAssetSet,
) -> dict[str, Any]:
    validate_rule_set(rule_set)
    if rule_set.status != "PUBLISHED":
        raise BrandRuleError("BrandContext requires a PUBLISHED BrandRuleSet")
    if token_set.brand_profile_id != rule_set.brand_profile_id:
        raise BrandRuleError("BrandContext token profile mismatch")
    if asset_set.brand_profile_id != rule_set.brand_profile_id:
        raise BrandRuleError("BrandContext asset profile mismatch")
    if token_set.version != rule_set.token_set_version:
        raise BrandRuleError("BrandContext token version mismatch")
    if asset_set.version != rule_set.asset_set_version:
        raise BrandRuleError("BrandContext asset version mismatch")

    hard_rules = [
        rule
        for rule in sorted(rule_set.rules, key=lambda item: (-item.priority, item.id))
        if rule.active and rule.severity == "HARD"
    ]
    allowed_assets = sorted(
        set(asset_set.logo_asset_ids)
        | set(asset_set.font_asset_ids)
        | set(asset_set.reference_asset_ids)
    )
    return {
        "brand_profile_id": rule_set.brand_profile_id,
        "brand_rule_set_id": rule_set.id,
        "brand_rule_set_version": rule_set.version,
        "hard_rule_ids": [rule.id for rule in hard_rules],
        "selected_tokens": {
            "colors": dict(sorted(token_set.colors.items())),
            "font_asset_ids": list(token_set.font_asset_ids),
            "spacing_scale": list(token_set.spacing_scale),
        },
        "allowed_assets": allowed_assets,
        "voice_summary": dict(rule_set.voice),
        "reference_asset_ids": sorted(asset_set.reference_asset_ids),
        "pinned": True,
    }
